File: sdk/java/src/test/java/jp/kengos/shojiku/TemplateRootTest.java
package jp.kengos.shojiku;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Comparator;
import java.util.Map;
import java.util.stream.Stream;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.ValueSource;

/**
 * The template-root hardening, one test per claim.
 *
 * <p>A hostile NAME is a fact about the request, so every case here is a FAILED RESULT rather than
 * an exception — the distinction a caller branches on. The rules are the UNION across platforms, so
 * a name refused on Linux is refused on Windows and the same application deploys to both.
 */
class TemplateRootTest extends Fixtures {

  @ParameterizedTest
  @ValueSource(
      strings = {
        // Blank: a hostile STRING, not misuse — it can arrive straight from a form field.
        "",
        "   ",
        // Traversal, in both separators, and what subsumes it: a name is one segment.
        "../receipt",
        "..\\receipt",
        "nested/receipt",
        "nested\\receipt",
        // Absolute, POSIX and Windows.
        "/etc/passwd",
        "\\\\host\\share",
        // Drive-relative: Windows resolves this against that drive's current directory.
        "C:receipt",
        // Control characters, including the NUL a C boundary must never be handed.
        "recei\0pt",
        "recei\npt",
        "recei\u001bpt",
        // Reserved DOS devices, including the trailing dots and spaces Windows strips first.
        "CON",
        "nul",
        "CON.",
        "CON ",
        "LPT1",
        "COM9.txt"
      })
  void aHostileNameIsARefusedRequest(String name) {
    Result<DocumentArtifact> result = client().build().generate(name, null);

    assertTrue(result.failed(), name);
    assertEquals("template_name", result.failure().kind());
    assertEquals(Step.GENERATE, result.failure().step());
  }

  @Test
  void aNullNameIsARefusedRequestToo() {
    // It can arrive from a deserialized payload with a missing field, which is
    // the same class of thing as a blank one: a fact about the request.
    Result<DocumentArtifact> result = client().build().generate(null, null);

    assertTrue(result.failed());
    assertEquals("template_name", result.failure().kind());
  }

  @Test
  void aRefusalCapsHowMuchOfTheNameItEchoes() {
    // A name reaches exception reporters and log files, so the echo is bounded —
    // the same discipline the engine applies to what it echoes.
    String hostile = "C:" + "x".repeat(500);

    Result<DocumentArtifact> result = client().build().generate(hostile, null);

    assertTrue(result.failed());
    assertFalse(result.failure().message().contains("x".repeat(200)));
  }

  @Test
  void aRefusalStripsControlCharactersOutOfTheEcho() {
    // A separate surface from the cap: a hostile name must not be able to smuggle
    // an escape sequence into a terminal or a log aggregator.
    Result<DocumentArtifact> result = client().build().generate("recei\u001bpt", null);

    assertTrue(result.failed());
    assertFalse(result.failure().message().contains("\u001b"));
  }

  @Test
  void aNameThatIsNotThereIsARefusedRequestNamingWhy() {
    Result<DocumentArtifact> result = client().build().generate("no_such_template", null);

    assertTrue(result.failed());
    assertEquals("template_not_found", result.failure().kind());
    // The underlying io error rides as the CAUSE, not as the headline.
    assertNotNull(result.failure().cause());
    assertEquals("io", result.failure().cause().kind());
    assertEquals(2, result.failure().causes().size());
  }

  @Test
  void aSymlinkPointingOutsideTheRootIsNotFollowed() throws IOException {
    // The check a name-shape rule cannot make: this name passes every rule above
    // and still points out.
    Path outside = Files.createTempDirectory("shojiku-outside");
    Path root = Files.createTempDirectory("shojiku-root");
    try {
      Files.copy(
          Paths.get(TEMPLATES, "receipt", "templates.yml"), outside.resolve("templates.yml"));
      Files.createSymbolicLink(root.resolve("escape"), outside);

      Result<DocumentArtifact> result =
          client().templates(root.toString()).build().generate("escape", null);

      assertTrue(result.failed());
      assertEquals("template_escapes_root", result.failure().kind());
    } finally {
      delete(root);
      delete(outside);
    }
  }

  @Test
  void aSymlinkStayingInsideTheRootIsFine() throws IOException {
    // Containment is about where the answer LANDS, not about symlinks.
    Path root = Files.createTempDirectory("shojiku-root");
    try {
      Path real = Files.createDirectory(root.resolve("real"));
      Files.copy(Paths.get(TEMPLATES, "receipt", "templates.yml"), real.resolve("templates.yml"));
      Files.createSymbolicLink(root.resolve("alias"), real);

      assertTrue(client().templates(root.toString()).build().generate("alias", null).success());
    } finally {
      delete(root);
    }
  }

  @Test
  void aSiblingDirectoryWithTheRootsNameAsAPrefixIsNotInsideIt() throws IOException {
    // What a string prefix compare gets wrong: `/tmp/root-evil` starts with
    // `/tmp/root`. The containment test is structural instead.
    Path parent = Files.createTempDirectory("shojiku-prefix");
    try {
      Path root = Files.createDirectory(parent.resolve("root"));
      Path evil = Files.createDirectories(parent.resolve("root-evil").resolve("receipt"));
      Files.copy(Paths.get(TEMPLATES, "receipt", "templates.yml"), evil.resolve("templates.yml"));
      Files.createSymbolicLink(root.resolve("receipt"), evil);

      Result<DocumentArtifact> result =
          client().templates(root.toString()).build().generate("receipt", null);

      assertTrue(result.failed());
      assertEquals("template_escapes_root", result.failure().kind());
    } finally {
      delete(parent);
    }
  }

  @Test
  void anUnreadableTemplateIsARefusedRequest() throws IOException {
    // Structurally unreadable, not `chmod 000`: the gate container runs as root,
    // which ignores permission bits — that test would pass for the wrong reason.
    // A directory where the file belongs cannot be read by anyone.
    Path root = Files.createTempDirectory("shojiku-unreadable");
    try {
      Files.createDirectories(root.resolve("shadow").resolve("templates.yml"));

      Result<DocumentArtifact> result =
          client().templates(root.toString()).build().generate("shadow", null);

      assertTrue(result.failed());
      assertEquals("template_unreadable", result.failure().kind());
      assertEquals("io", result.failure().cause().kind());
    } finally {
      delete(root);
    }
  }

  @Test
  void definitionsAreOptional() {
    // `warns` has no definitions.yml; `receipt` has one. Both resolve.
    assertTrue(client().build().generate("warns", null).success());
    assertTrue(client().build().generate("receipt", null).success());
  }

  @Test
  void theEnvironmentSuppliesTheRoot() {
    // Through the rule itself, with the environment injected: a Java process
    // cannot set a variable in its own environment, so a test that went through a
    // built client could only ever prove the explicit half of the precedence.
    Env env = new Env(true, Map.of("SHOJIKU_TEMPLATE_ROOT", TEMPLATES));

    assertEquals(TEMPLATES, Settings.resolveRoot(null, env));
  }

  @Test
  void explicitConfigurationBeatsTheEnvironment() {
    // The deliberate asymmetry with the LIBRARY: what an application renders is
    // the application's own decision, so the explicit value wins here — while
    // SHOJIKU_LIBRARY beats an explicit library path, which LibraryTest proves
    // end to end against the variable the gate image really sets.
    Env env = new Env(true, Map.of("SHOJIKU_TEMPLATE_ROOT", TEMPLATES));

    assertEquals("/explicit", Settings.resolveRoot("/explicit", env));
  }

  @Test
  void theEnvironmentKnobDisablesTheLookup() {
    Env silenced = new Env(false, Map.of("SHOJIKU_TEMPLATE_ROOT", TEMPLATES));

    assertNull(Settings.resolveRoot(null, silenced));
    // And end to end: a client built with env(false) and no explicit root has none.
    assertNull(rootless().build().templateRootOrNull());
  }

  @Test
  void nothingSuppliesARootWhenNeitherPositionDoes() {
    assertNull(Settings.resolveRoot(null, new Env(true, Map.of())));
  }

  private static void delete(Path directory) throws IOException {
    try (Stream<Path> walk = Files.walk(directory)) {
      walk.sorted(Comparator.reverseOrder()).map(Path::toFile).forEach(java.io.File::delete);
    }
  }
}

File: sdk/js/test/templateRoot.test.ts
/**
 * Template names are identifiers, not paths — one test per refused class.
 *
 * The rejection rules are the UNION across platforms, not this host's. Windows
 * is a first-class target, so a backslash is a separator, `C:name` is
 * drive-relative, `\\host\share` is a UNC path and `CON`/`NUL` are reserved
 * devices — every one of them refused on EVERY platform.
 *
 * Every refusal is a FAILED RESULT, never a throw: a hostile name is a fact
 * about the request, not a bug in the calling program.
 */

import { mkdir, mkdtemp, symlink, writeFile } from 'node:fs/promises';
import { tmpdir } from 'node:os';
import { join } from 'node:path';
import { afterEach, describe, expect, it } from 'vitest';
import { ECHO_LIMIT } from '../src/errors.js';
import { resetConfiguration } from '../src/index.js';
import { FIXTURE_TEMPLATES, makeClient } from './support/fixtures.js';

afterEach(resetConfiguration);

async function refused(name: string) {
  return makeClient().generate(name, {});
}

describe('refused name classes', () => {
  it('refuses an empty name', async () => {
    const result = await refused('');
    expect(result.failed).toBe(true);
    expect(result.failure?.kind).toBe('template_name');
  });

  it('refuses a blank-but-not-empty name', async () => {
    expect((await refused('   ')).failure?.kind).toBe('template_name');
  });

  it('refuses an absolute path', async () => {
    expect((await refused('/etc/passwd')).failure?.kind).toBe('template_name');
  });

  it('refuses a drive-relative name, which Windows resolves per drive', async () => {
    expect((await refused('C:receipt')).failure?.kind).toBe('template_name');
  });

  it('refuses a UNC path', async () => {
    expect((await refused('\\\\host\\share')).failure?.kind).toBe('template_name');
  });

  it('refuses `..` traversal', async () => {
    expect((await refused('../receipt')).failure?.kind).toBe('template_name');
  });

  it('refuses a forward separator', async () => {
    expect((await refused('nested/receipt')).failure?.kind).toBe('template_name');
  });

  it('refuses a backslash separator, on every platform', async () => {
    expect((await refused('nested\\receipt')).failure?.kind).toBe('template_name');
  });

  it('refuses a control character', async () => {
    expect((await refused('recei\x00pt')).failure?.kind).toBe('template_name');
    expect((await refused('recei\x1bpt')).failure?.kind).toBe('template_name');
  });

  it('refuses reserved DOS device names, with or without an extension', async () => {
    expect((await refused('CON')).failure?.kind).toBe('template_name');
    expect((await refused('nul')).failure?.kind).toBe('template_name');
    expect((await refused('COM1.yml')).failure?.kind).toBe('template_name');
    // Windows STRIPS trailing dots and spaces before resolving, so these are
    // the CON device too.
    expect((await refused('CON.')).failure?.kind).toBe('template_name');
    expect((await refused('CON ')).failure?.kind).toBe('template_name');
  });

  it('bounds and strips the name it echoes back', async () => {
    const hostile = `${'a'.repeat(200)}\x07/x`;
    const message = (await refused(hostile)).failure?.message ?? '';

    // The invariant is about the ECHOED TEXT, not the whole message: the
    // explanation beside it is longer than some hostile names, so comparing
    // total lengths would measure the explanation instead of the cap.
    expect(message).not.toContain('\x07');
    expect(/a+/.exec(message)?.[0].length ?? 0).toBeLessThanOrEqual(ECHO_LIMIT);
  });

  it('does not resolve a prototype name through an inherited property', async () => {
    // A plain-object lookup table would answer `constructor` with an inherited
    // FUNCTION, which reads as a hit. These must all be ordinary misses.
    for (const name of ['constructor', '__proto__', 'toString']) {
      const result = await refused(name);
      expect(result.failed).toBe(true);
      expect(result.failure?.kind).toBe('template_not_found');
    }
  });
});

describe('containment', () => {
  it('reports a name that is shaped fine but absent', async () => {
    expect((await refused('nosuchtemplate')).failure?.kind).toBe('template_not_found');
  });

  it('carries the underlying cause under the refusal', async () => {
    const failure = (await refused('nosuchtemplate')).failure;

    expect(failure?.cause?.kind).toBe('io');
    // Outermost first, so a log shows the refusal before its reason.
    expect(failure?.causes.map((item) => item.kind)).toEqual(['template_not_found', 'io']);
  });

  it('does NOT follow a symlink that points outside the root', async () => {
    const base = await mkdtemp(join(tmpdir(), 'shojiku-root-'));
    const root = join(base, 'root');
    const outside = join(base, 'outside');
    await mkdir(root, { recursive: true });
    await mkdir(outside, { recursive: true });
    await writeFile(join(outside, 'templates.yml'), 'version: 0.1.0\n');
    await symlink(outside, join(root, 'escape'));

    const result = await makeClient({ templates: root }).generate('escape', {});

    // The check a name-shape rule cannot make: after following the filesystem,
    // is the answer still inside the root?
    expect(result.failed).toBe(true);
    expect(result.failure?.kind).toBe('template_escapes_root');
  });

  it('is structural, so a sibling directory sharing the root’s prefix cannot pass', async () => {
    const base = await mkdtemp(join(tmpdir(), 'shojiku-prefix-'));
    const root = join(base, 'root');
    await mkdir(root, { recursive: true });
    await mkdir(join(base, 'root-evil'), { recursive: true });
    await writeFile(join(base, 'root-evil', 'templates.yml'), 'version: 0.1.0\n');
    await symlink(join(base, 'root-evil'), join(root, 'sneak'));

    const result = await makeClient({ templates: root }).generate('sneak', {});

    // A string prefix compare would admit `/…/root-evil` as inside `/…/root`.
    expect(result.failure?.kind).toBe('template_escapes_root');
  });

  it('reads the optional definitions file when there is one, and copes when there is not', async () => {
    const withDefinitions = await makeClient().generate('receipt', {
      customer: { name: 'x' },
    });
    const without = await makeClient().generate('warns', {});

    expect(withDefinitions.success).toBe(true);
    expect(without.success).toBe(true);
  });

  it('refuses a definitions.yml that is there and cannot be read', async () => {
    const base = await mkdtemp(join(tmpdir(), 'shojiku-defs-'));
    const entry = join(base, 'has-bad-definitions');
    await mkdir(entry, { recursive: true });
    await writeFile(join(entry, 'templates.yml'), 'version: 0.1.0\n');
    // A DIRECTORY where the definitions file belongs: unreadable structurally,
    // which holds even in a container running as root where a chmod would not.
    await mkdir(join(entry, 'definitions.yml'), { recursive: true });

    const result = await makeClient({ templates: base }).generate('has-bad-definitions', {});

    // ABSENCE is what makes definitions optional, not unreadability: swallowing
    // this would render the document as though it declared no schema at all.
    expect(result.failure?.kind).toBe('template_unreadable');
  });

  it('refuses a template directory whose templates.yml cannot be read', async () => {
    const base = await mkdtemp(join(tmpdir(), 'shojiku-unreadable-'));
    // A DIRECTORY where the template file belongs: structurally unreadable, so
    // it holds even in a container running as root, where a chmod would not.
    await mkdir(join(base, 'broken-entry', 'templates.yml'), { recursive: true });

    const result = await makeClient({ templates: base }).generate('broken-entry', {});

    expect(result.failure?.kind).toBe('template_unreadable');
  });

  it('resolves the ordinary case against the configured root', async () => {
    expect(makeClient().templateRoot?.path).toBe(FIXTURE_TEMPLATES);
  });
});
